### ISSUE TITLE
Improve collection item private metadata

### ISSUE THREAD

**Issue by DavidWells** [MoOx/phenomic#712]
I'm trying to trim down the window.collect as much as possible. It would appear that some of the values in the default object have very similar values and I'm wondering if we might be able to trim them down.

https://github.com/MoOx/phenomic/blob/1a11c7f8db4bdc373b72764535f6e5629e042f0a/src/phenomic-loader/minify.js#L13-L16

For example: `__resourceUrl` and `__url` are almost always the same except for the ending `index.html` on resourceURL.

```
    __filename: item.__filename,
    __url: item.__url,
    __resourceUrl: item.__resourceUrl,
    __dataUrl: item.__dataUrl,
```

Perhaps there is a way to minimize this further.

Just a thought, trying to understand/minimize stuff for large sites again =)

**Comment by thangngoc89** [comment]
You can optimize __dataUrl field. 

Currently, the *.json needed for loading the page will be place next to the index.html file.

This adds extra bytes on window.collection if the path is too long. Since we track the dataUrl manually, we can put all the *.json files into a folder. Say phenomic/data and only specify the file name then. @MoOx what do you think? 

By the way, I'm working on a big site too.

**Comment by MoOx** [comment]
Maybe we can improve that indeed. Instead of the full value in resourceUrl, we might just add the diff that should be added (or not) from `__url`.

**Comment by thangngoc89** [comment]
@MoOx could you explain the different between resource and url? A full implementation if diffing algorithm will add a lot of bytes to the main bundle. So if I can see the different between between url and resource, I can make the diffing shorter

**Comment by MoOx** [comment]
url: /blah/
resourceUrl: /blah/index.html

When I say diff, I mean resourceUri: index.html or ""

**Comment by thangngoc89** [comment]
So the resourceUrl always url + string right? No other cases. We can omit resourceUrl field if url === resourceUrl

**Comment by MoOx** [comment]
Yeah probably. Will just need to adjust code that use resourceUrl.

**Comment by thangngoc89** [comment]
What about my idea above? Moving all .json to a folder. It could save some bytes without big changes

**Comment by MoOx** [comment]
Why not.

**Comment by thangngoc89** [comment]
Also, you shorter hash. We dont need a full hash here.

**Comment by thangngoc89** [comment]
This is the code that generate `url` and `resourceUrl`

```
  const url = urlify(tmpUrl)
  const resourceUrl = urlify(tmpUrl, true)
```

@MoOx instead of diffing and do some weird trick to generate resourceUrl, can we bundle the `urlify` utility in client ? Might need webpack 2 those. Because it requires `path.join`

**Comment by thangngoc89** [comment]
@DavidWells @MoOx I did some optimization based on our discussion here. 

# Commands 

- Calculate size of all html files inside dist folder

```
$ du -ch -- dist/**/*.html | tail -n 1
```

- Count number of files. Make sure we didn't mess up anything

```
$ du -ch -- dist/**/*.html | wc -l
56
```

```
$ du -ch -- dist/**/*.html | wc -l
45
```

# Results

- Initial: 3,1M
- Make digestHash shorter (5 chars): 3,0M
- Use a central folder for storing .json file: 2,9M
- Remove __resourceUrl field: 2,8M

# Discussion

The last one (Remove __resourceUrl) requires many changes so I'm considering doesn't worth it. 
@DavidWells If you have a chance, could you do a step-by-step test like this for me on your side? 

I pushed my work here https://github.com/MoOx/phenomic/tree/optimize-collection
The commit message pretty much the same with the step name above. Thanks

**Comment by thangngoc89** [comment]
This is a quick test on a site with 64 html files. Looks like removing __resourceUrl is the biggest improvement here

```
- Initial: 4,7M
- Make digestHash shorter (5 chars): 4,6M
- Use a central folder for storing .json file: 4,5M
- Remove __resourceUrl field: 4,2M
```

**Comment by MoOx** [comment]
The best improvement will be to stop shipping the entire collection, so work on #713

**Comment by bloodyowl** [comment]
Just an idea: when passing props with the `PhenomicContainer` is defining ES5 getters instead of regular properties so that we can ship exactly what's needed. Bit risky considering some components can only render on the client, but that can be a nice way to keep the file size down.

**Comment by MoOx** [comment]
Not sure why getters will help for this as we might statically call the method to get items before prerendering the page (so before injecting data in the html).
Oh you mean using getters on props of each items. Clever (but risky indeed)!
You are smart.

**Comment by bloodyowl** [comment]
Going to be fixed with #925

**Comment by MoOx** [comment]
v1 have a new way to load data.
